src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/ExtractToCssCodeActionProvider.cs

```diff
@@ -0,0 +1,95 @@
+    {
+        if (!context.SupportsFileCreation)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!context.CodeDocument.FileKind.IsComponent())
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!context.CodeDocument.TryGetSyntaxRoot(out var root))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var owner = root.FindInnermostNode(context.StartAbsoluteIndex);
+        if (owner is null)
+        {
+            _logger.LogWarning($"Owner should never be null.");
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (owner is MarkupTextLiteralSyntax { Parent: MarkupStartTagSyntax })
+        {
+            owner = owner.Parent;
+        }
+
+        // We have to be in a style tag (or inside it, but we'll have moved to the parent if so, above)
+        if (owner is not (MarkupStartTagSyntax { Name.Content: "style" } or MarkupEndTagSyntax { Name.Content: "style" }))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // If there is any C# or Razor in the style tag, we can't offer, so it has to be one big text literal.
+        if (owner.Parent is not MarkupElementSyntax { Body: [MarkupTextLiteralSyntax textLiteral] } markupElement ||
+            textLiteral.ChildNodes().Count() > 0)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // If there are diagnostics, we can't trust the tree to be what we expect.
+        if (markupElement.GetDiagnostics().Any(d => d.Severity == RazorDiagnosticSeverity.Error))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var actionParams = new ExtractToCssCodeActionParams()
+        {
+            ExtractStart = textLiteral.Span.Start,
+            ExtractEnd = textLiteral.Span.End,
+            RemoveStart = markupElement.Span.Start,
+            RemoveEnd = markupElement.Span.End
+        };
+
+        var resolutionParams = new RazorCodeActionResolutionParams()
+        {
+            TextDocument = context.Request.TextDocument,
+            Action = LanguageServerConstants.CodeActions.ExtractToCssAction,
+            Language = RazorLanguageKind.Razor,
+            DelegatedDocumentUri = context.DelegatedDocumentUri,
+            Data = actionParams,
+        };
+
+        var razorFileName = Path.GetFileName(context.Request.TextDocument.DocumentUri.GetAbsoluteOrUNCPath());
+
+        var codeAction = RazorCodeActionFactory.CreateExtractToCss(razorFileName, resolutionParams);
+        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
```

Comment:
super-nit: this line spacing looks a bit funky to me. Seems like all three lines should be together, or have blank line between each one.